Paraswap-Agents/delegate-calls/src/agent.ts

```diff
@@ -0,0 +1,96 @@
+
+import LRU from "lru-cache";
+
+import { createFinding } from "./finding";
+
+const cache: LRU<string, boolean> = new LRU<string, boolean>({ max: 10000 });
+
+// The address of the Paraswap contract (the contract's name is "AugustusSwapper")
+const AUGUSTUS_ADDR = "0xdef171fe48cf0115b1d80b88dc8eab59176fee57";
+
+// bytes32 representation of the `ROUTER_ROLE` signature (`"ROUTER_ROLE` == "0x524f555445525f524f4c45`)
+const ROUTER_ROLE = ethers.utils.keccak256("0x524f555445525f524f4c45");
+
+export const PARA_ABI = [
+  "function hasRole(bytes32 role, address account) public view returns (bool)",
+  "event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)",
+  "event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)",
+];
+
+// Checks if `logicAddr` has the role `ROUTER_ROLE` in the contract `paraAddr`
+const checkRouterRole = async (paraAddr: string, logicAddr: string): Promise<boolean | undefined> => {
+  let isRouterRole;
+  // If the contract exists in the cache then return data
+  if (cache.has(logicAddr)) {
+    isRouterRole = cache.get(logicAddr);
+  // Otherwise the cache does not contain data for `logicAddr`
+  } else {
+    try {
+      // Check on-chain data to see if `logicAddr` has the correct role
+      const contractInterface = new ethers.Contract(paraAddr, PARA_ABI, getEthersProvider());
+      isRouterRole = await contractInterface.hasRole(ROUTER_ROLE, logicAddr);
+      cache.set(logicAddr, isRouterRole);
+    } catch (error) {
+      // If there is as error set `isRouterRole` to undefined
+      // A special finding to indicate that the on-chain call failed is generated when `isRouterRole` is undefined
+      // This is not saved to cache in case it was a short term issue
+      isRouterRole = undefined;
+    }
+  }
+  return isRouterRole;
+};
+
+export const provideHandleTransaction = (address: string, checkRouterRole: any): HandleTransaction => {
+  return async (tx: TransactionEvent): Promise<Finding[]> => {
+    // Setup the findings array
+    const findings: Finding[] = [];
+
+    // Check if any contracts have had roles granted and update `cache` if needed
+    const roleGrants = tx.filterLog(PARA_ABI[1], address);
+    roleGrants.forEach((roleGrant) => {
+      // If ROUTER_ROLE was granted to an account that is inside the cache then update the cache
+      if (roleGrant.args.role == ROUTER_ROLE && cache.has(roleGrant.args.account)) {
+        cache.set(roleGrant.args.account, true);
+      }
+    });
+
+    // Check if any contracts have had roles revoked and update `cache` if needed
+    const roleRevokes = tx.filterLog(PARA_ABI[2], address);
+    roleRevokes.forEach((roleRevoke) => {
+      // If ROUTER_ROLE was removed from an account that is inside the cache then update the cache
+      if (roleRevoke.args.role == ROUTER_ROLE && cache.has(roleRevoke.args.account)) {
+        cache.set(roleRevoke.args.account, false);
+      }
+    });
```

Comment:
Can't we avoid these updates with every tx by specifying the block number in `hasRole` call? 
it'll cause updates to be done with each blockEvent

Reply:
I've added specifying the block number to the call to `hasRole` in the function `checkRouterRole` to prevent web3 calls to a future block while the agent is still working on the previous block.

As for you suggestion to change the way we keep up to date with role changes, the current implementation and your suggestion both have positives and negatives.

New suggestion:
- Removes need to run `tx.filterLog` for on role grant and revoke events
- Adds need to make a web3 call to `AugustusSwapper.hasRole` on every delegated call

Current implementation:
- Needs to run `tx.filterLog` twice
- Number of web3 calls are reduced since the roles and their associated addresses are stored in the LRU cache.

@e1Ru1o 
Yasmine and I discussed this in a call and aren't sure which approach is better. Thoughts?

Reply:
I understand what you are doing and it seems like a great effort to improve performance. I guess that you are right and your implementation has a better performance because the network calls are more expensive than the filters.

So my only concern here is that if for some reason the agent miss a txn or analyze some txns out of order I guess false positive/negative can happen with the data you have in the cache since it will not be updated correctly. In other words your approach is dependent of the order in where the txns are processed. Forta team tell us before that this is not a problem and we can assume the txns are going to be processed in order, but we always try to be careful with that due to the asynchronous behavior of the agents. 

I just checked the sdk [here](https://github.com/forta-protocol/forta-agent-sdk/blob/master/cli/commands/run/run.live.ts#L8) and as Forta said we don't need to worry about that, blocks are being processed one by one as well as txns. 

Said the above I think you can keep your approach. You are setting the address to `true` or `false` accordingly to the events and in the cases you don't have info in the cache you are fetching the data from the blockchain so everything looks good to me right now.